packages/context/src/handler-chain.ts

```diff
@@ -0,0 +1,210 @@
+// Copyright IBM Corp. 2019. All Rights Reserved.
+// Node module: @loopback/context
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import * as debugFactory from 'debug';
+import {BindingFilter} from './binding-filter';
+import {BindingAddress} from './binding-key';
+import {BindingComparator} from './binding-sorter';
+import {Context} from './context';
+import {transformValueOrPromise, ValueOrPromise} from './value-promise';
+
+const debug = debugFactory('loopback:context:handler-chain');
+
+/**
+ * Return value for a method invocation
+ */
+// tslint:disable-next-line:no-any
+export type InvocationResult = any;
+
+/**
+ * Array of arguments for a method invocation
+ */
+// tslint:disable-next-line:no-any
+export type InvocationArgs = any[];
+
+/**
+ * `next` function
+ */
+export type Next = () => ValueOrPromise<InvocationResult>;
+
+/**
+ * A handler function to be invoked in a chain for the given context
+ */
+export interface Handler<C extends Context = Context> {
+  /**
+   * @param context - Context object
+   * @param next - A function to proceed with downstream handlers or the target
+   *
+   * @returns The invocation result as a value or promise
+   */
+  (context: C, next: Next): ValueOrPromise<InvocationResult>;
```

Comment:
I am not a fan of defining interfaces for function types, I find such type more difficult to understand. I think it's even more confusing to novice TypeScript users. 

Another aspect that I don't like is that you have two TSDoc comments now - one for the interface, another for the actual function.

What are you arguments for preferring interfaces?

My assumption is that the following syntax should work equally well (if not better), please correct me if I am wrong and/or missing something.

```ts
/**
 * A handler function to be invoked in a chain for the given context
 * @param context - Context object
 * @param next - A function to proceed with downstream handlers or the target
 *
 * @returns The invocation result as a value or promise
 */
export type Handler<C extends Context = Context> = 
  (context: C, next: Next) => ValueOrPromise<InvocationResult>;
```

Reply:
I'm not sure if a `type` definition can have `@param` and `@returns`. That's why I use the interface style.

Reply:
I switched to `type`.